### ISSUE TITLE
VeRA doesn't work with lower-precision models

### ISSUE THREAD

**Issue by 51616** [huggingface/peft#1872]
### System Info

python 3.10.14
accelerate 0.31.0
peft 0.11.2.dev0 (current main)
transformers  4.41.2

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
from peft import PeftModel, VeraConfig


device = "cuda"
dtype = torch.bfloat16 # or torch.float16

vera_config = VeraConfig(
    r=8,
    vera_dropout=0.05,
    target_modules=["q_proj", "v_proj"],
)


model = AutoModelForCausalLM.from_pretrained(
    "mistralai/Mistral-7B-v0.1", torch_dtype=dtype, device_map=device
)
model = PeftModel(model, vera_config)
model.set_adapter("default")

tokenizer = AutoTokenizer.from_pretrained("mistralai/Mistral-7B-v0.1")

model.eval()
tokenizer.padding_side = "left"
tokenizer.pad_token = tokenizer.eos_token

prompts = ["My mother", "My favourite", "I like to eat"]
model_inputs = tokenizer(prompts, return_tensors="pt", padding=True).to(device)

generated_ids = model.generate(**model_inputs, max_new_tokens=100, do_sample=True)
```

Error msg:
```
File "[REDACTED]/peft/tuners/vera/layer.py", line 267, in forward
    result = result + lambda_b * F.linear(lambda_d * F.linear(dropout(x), sliced_A), sliced_B)
RuntimeError: expected mat1 and mat2 to have the same dtype, but got: float != c10::Half
```
The error happens with either `torch.bfloat16` or `torch.float16`

Changing `device` to `torch.float32` doesn't have this problem.

A small patch in `peft/tuners/vera/layer.py`, casting `sliced_A` and `sliced_B` to the same dtype as `lambda_d` (float32) does workaround this problem:
```python
    def forward(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
        previous_dtype = x.dtype

        if self.disable_adapters:
            if self.merged:
                self.unmerge()
            result = self.base_layer(x, *args, **kwargs)
        elif self.merged:
            result = self.base_layer(x, *args, **kwargs)
        else:
            result = self.base_layer(x, *args, **kwargs)
            for active_adapter in self.active_adapters:
                if active_adapter not in self.vera_lambda_d.keys():
                    continue

                lambda_d = self.vera_lambda_d[active_adapter]
                lambda_b = self.vera_lambda_b[active_adapter]

                vera_A = self.vera_A[active_adapter]
                vera_B = self.vera_B[active_adapter]

                # As adapted layers may have different shapes and VeRA contains a single shared pair of A and B matrices,
                # we initialize these matrices with the largest required size for each dimension.
                # During the forward pass, required submatrices are sliced out from the shared vera_A and vera_B.
                sliced_A = vera_A[:, : self.in_features]
                sliced_B = vera_B[: self.out_features, :]
                
                dropout = self.vera_dropout[active_adapter]
                x = x.to(lambda_d.dtype)
                sliced_A = sliced_A.to(lambda_d.dtype)
                sliced_B = sliced_B.to(lambda_d.dtype)
                result = result + lambda_b * F.linear(lambda_d * F.linear(dropout(x), sliced_A), sliced_B)

        result = result.to(previous_dtype)
        return result
```

### Expected behavior

It shouldn't break when using lower-precision models, though I only tested with `mistralai/Mistral-7B-v0.1`.

I'm not sure why `vera_A` and `vera_B` are casted to `bfloat16` to begin with, they are initialized as `float32`. I dug around but can't find where they are casted to `bfloat16`.

Also, the patch is probably suboptimal since if we have to cast them to `float32` in every forward pass instead of setting them as `float32` and keep it that way.

**Comment by BenjaminBossan** [comment]
Thanks for reporting this issue. As you correctly stated, VeRA should work even if the model is loaded in float16 or bfloat16. 

I assume you're working with the latest main branch of PEFT. In that case, the casting to float32 happens because we want to keep the learnable parameters at high precision, as otherwise we might see training instability (see #1706 for more context).

To check this, try loading the model with `autocast_adapter_dtype=False`:

```diff
- model = PeftModel(model, vera_config)
+ model = get_peft_model(model, vera_config, autocast_adapter_dtype=False)
```

Since this is desired behavior, we should cast the adapter parameters to the correct dtype during `forward` as your snippet suggests. Are you interested in creating a PR to fix this?

**Comment by 51616** [comment]
Thank you for a quick response @BenjaminBossan.

I suppose what's happening here is that `vera_d` and `vera_b` are being casted "down" somewhere else rather than being cased up to `float32`. The patch I proposed casts them up from a lower precision to `float32` as both `lambda`s are already in `float32`.

I think main issue is that `vera_d` and `vera_b` are initialized as `float32` but being casted down somewhere after init. I didn't dig to the base model implementation but my guess is that `vera_A` and `vera_b` are not being casted to `float32` despite `autocast_adapter_dtype=True`.

**Comment by BenjaminBossan** [comment]
@51616 I created a PR that should fix this issue: #1874. If you have time, it would be great if you could check if this solves your particular problem.
